### PR TITLE
MXFP4 Cast Transpose Triton and Quantizer Infra

### PR DESCRIPTION
**Pull request by sarthak-amd**
# Description
Implements the MXFP4 `rowwise` and `columnwise` FP32/BF16 -> MXFP4  fused quantization + cast kernel

- [x] Verify Tolerances and functional Unit Tests

- The triton `te_cast_transpose_mxfp4_triton` currently outputs FP4 data in linear layout [M, N/2] with contiguous byte packing. AITER's `gemm_a4w4` requires the B matrix in MFMA shuffle layout for tensor cores. This layout shuffle can be fused into the triton kernel in future.

### CODE REVIEW

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py [hunk 1]
```diff
@@ -0,0 +1,178 @@
+# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
+# License for AMD contributions = MIT. See LICENSE for more information
+
+import math
+import pytest
+import torch
+import numpy as np
+import os
+
+os.environ["USE_TRITON_FUSED_CAST_TRANSPOSE"] = "1"
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_cast_mxfp4.py]: We previously already defined env NVTE_USE_CAST_TRANSPOSE_TRITON.

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+@pytest.mark.parametrize(("rowwise", "columnwise"), [
+    (True, True), 
+    (False, True), 
+    (True, False)
+])
+@pytest.mark.parametrize("shuffle_B_matrix", [False, True])
+def test_quantize_mxfp4(shape, in_dtype, rowwise, columnwise, shuffle_B_matrix):
+    """Test MXFP4 quantization for rowwise/columnwise modes with/without FP4 shuffle.
+    
+    Note: FP4 data shuffle (shuffle_B_matrix_for_aiter) is not yet supported in Triton kernel.
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_cast_mxfp4.py]: If FP4 data shuffle is not yet supported in Triton kernel, why do we need to add it here?

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+@pytest.mark.parametrize("shape", [
+    (128, 128),
+    (256, 256),
+    (256, 1024),
+    (2048, 6144),
+    (16384, 128),
+    (32768, 160),
+    (4096, 1632),
+    (8, 32, 1024),
+    (16, 8, 4, 512),
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_cast_mxfp4.py]: Can we add some prime numbers like https://github.com/ROCm/TransformerEngine/blob/9d6b0e55d5affbdad34352d697032dd67419ee19/tests/cpp/operator/test_cast_transpose.cu#L90-L92

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+        rowwise=rowwise, 
+        columnwise=columnwise,
+        shuffle_B_matrix_for_aiter=shuffle_B_matrix
+    )
+    out = quantizer.make_empty(input_tensor.shape, dtype=in_dtype)
+    quantized_out = te_quantize_triton(input_tensor, quantizer=quantizer, output=out)
+
+    # Tolerance: allow 1 nibble diff for rare edge cases near FP4 boundaries
+    data_atol = 20.0 if in_dtype != torch.float32 else 16.0
+    scale_atol = 2.0 if in_dtype != torch.float32 else 1.0
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_cast_mxfp4.py]: Data tol seems to be quite large. You can follow our mxfp8 scale and data adjustment scheme: https://github.com/ROCm/TransformerEngine/blob/9d6b0e55d5affbdad34352d697032dd67419ee19/tests/cpp/test_common.cu#L730

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        te_compare_results(
+            quantized_out._rowwise_data.view(torch.uint8),
+            ref_data,
+            atol=data_atol,
+            rtol=0.0,
+            msg="rowwise FP4 data mismatch",
+            use_torch_semantics=True
+        )
+        
+        # Compare only valid (non-padded) region - no shuffle extraction needed
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_cast_mxfp4.py]: What is fp4 shuffle?

File: transformer_engine/common/util/pybind_helper.h [hunk 6]
```diff
@@ -108,7 +108,8 @@
       .value("kFloat16", transformer_engine::DType::kFloat16)                                      \
       .value("kBFloat16", transformer_engine::DType::kBFloat16)                                    \
       .value("kFloat8E4M3", transformer_engine::DType::kFloat8E4M3)                                \
-      .value("kFloat8E5M2", transformer_engine::DType::kFloat8E5M2);                               \
+      .value("kFloat8E5M2", transformer_engine::DType::kFloat8E5M2)                                \
+      .value("kFloat4E2M1", transformer_engine::DType::kFloat4E2M1);                               \
```
Comment (wangye805) [inline transformer_engine/common/util/pybind_helper.h:112]: If we are going to enable kFloat4E2M1, there are other related changes needed. Search for https://github.com/search?q=repo%3AROCm%2FTransformerEngine%20kFloat4E2M1&type=code for more details: 

Image

File: transformer_engine/pytorch/tensor/storage/mxfp4_tensor_storage.py [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+    """Mixin class that holds data attributes of MXFP4Tensor.
+
+    MXFP4Tensor inherits from the PyTorch tensor class and this mixin
+    class. If this class is instantiated directly, it has the same
+    data, lower CPU overhead, and less functionality. It should only
+    be instantiated directly for performance-critical internal usage.
+
+    FP4 data format:
+    - Data: [M, K/2] uint8 tensor (2 FP4 values packed per byte)
+    - Scale: [M, K/32] uint8 tensor (E8M0 format, one scale per 32-element block)
```
Comment (wangye805) [inline transformer_engine/pytorch/tensor/storage/mxfp4_tensor_storage.py:62]: Is there alignment/padding requirements for M and K?

File: transformer_engine/pytorch/tensor/mxfp4_tensor.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+        # Update FP4 dtype
+        dst._fp4_dtype = self.dtype
+
+        return dst
+
+    def is_quantizable(self, inp: torch.Tensor) -> bool:
+        """Returns whether or not given inp can be quantized"""
+        if inp.ndim < 2:
+            return False
```
Comment (wangye805) [inline transformer_engine/pytorch/tensor/mxfp4_tensor.py:94]: TE currently supported 2D matrices from flatted high-dimensional tensors: https://github.com/ROCm/TransformerEngine/blob/9d6b0e55d5affbdad34352d697032dd67419ee19/transformer_engine/common/common.h#L238-L262

File: transformer_engine/pytorch/tensor/mxfp4_tensor.py [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+
+        def cdiv(a, b):
+            return (a + b - 1) // b
+
+        # Allocate FP4 data: [M, K/2]
+        rowwise_data = torch.empty(M, K // 2, dtype=torch.float4_e2m1fn_x2, device=device)
+        
+        # Allocate PADDED scale tensors for shuffle compatibility
+        rowwise_scale_N = K // MXFP4_BLOCK_SCALING_SIZE
+        rowwise_scale_M_pad = cdiv(M, 256) * 256
```
Comment (wangye805) [inline transformer_engine/pytorch/tensor/mxfp4_tensor.py]: I presume this 256 is from some alignment/padding requirement?

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
+# Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_cast_mxfp4.py:1]: You will need to add this pytest into our ci script (somewhere near https://github.com/ROCm/TransformerEngine/blob/9d6b0e55d5affbdad34352d697032dd67419ee19/ci/pytorch.sh#L74) otherwise it won't be tested

**Review by wangye805 — CHANGES_REQUESTED** [review]
You

File: transformer_engine/pytorch/tensor/mxfp4_tensor.py (hunk 9 above)
Comment (sarthak-amd) [reply transformer_engine/pytorch/tensor/mxfp4_tensor.py]: The 256 alignment is required by AITER's CK-based MXFP4 GEMM kernels for scale tensor swizzle/shuffle layout.
Required for scale swizzle layout: 256 = ScaleBlockSize(32) × 8 waves.
See `aiter/aiter/utility/fp4_utils.py:398 and gemm_a4w4_blockscale_common.cuh:66`

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py (hunk 5 above)
Comment (sarthak-amd) [reply tests/pytorch/triton_kernels/test_cast_mxfp4.py]: fp4 shuffle basically rearranges [M, K/2] linear layout → MFMA instruction layout (16×16).

The currently flow training workflow if TE MXFP4 Quantization Kernel is used is as follows
`TE Triton Kernel → Linear FP4 [N, K/2] → aiter.ops.shuffle_weight() → MFMA FP4 → aiter.gemm_a4w4()`

You can find the shuffle code in `aiter/aiter/ops/shuffle.py`

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py (hunk 2 above)
Comment (sarthak-amd) [reply tests/pytorch/triton_kernels/test_cast_mxfp4.py]: This is kept to ensure API consistency between Triton and the upcoming hip kernel for which I'll create a separate PR. In the hip kernel we were able to fuse the shuffle.

Comment (sarthak-amd) [reply tests/pytorch/triton_kernels/test_cast_mxfp4.py]: hip vs triton flow
Input: BF16 [M, N]
    ↓
MXFP4Quantizer.update_quantized()
    ↓
tex.cast_transpose_mxfp4_fused_shuffle() [Single HIP kernel]
    ↓
    ├─→ Rowwise FP4 [M, K/2] (MFMA shuffled)
    ├─→ Rowwise Scale [M_pad, K/32_pad] (shuffled)
    ├─→ Colwise FP4 [N, M/2] (MFMA shuffled)
    └─→ Colwise Scale [N_pad, M/32_pad] (shuffled)
    ↓
AITER gemm_a4w4 (zero-copy)

vs 

Input: BF16 [M, N]
    ↓
MXFP4Quantizer.update_quantized()
    ↓
te_cast_transpose_mxfp4_triton() [Triton JIT kernel]
    ↓
    ├─→ Rowwise FP4 [M, K/2] (linear layout)
    ├─→ Rowwise Scale [M_pad, K/32_pad] (shuffled)
    ├─→ Colwise FP4 [N, M/2] (linear layout)
    └─→ Colwise Scale [N_pad, M/32_pad] (shuffled)
    ↓
aiter.ops.shuffle.shuffle_weight() [External call]
    ↓
FP4 data → MFMA layout
    ↓
AITER gemm_a4w4

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py (hunk 10 above)
Comment (sarthak-amd) [reply tests/pytorch/triton_kernels/test_cast_mxfp4.py:1]: done

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py (hunk 3 above)
Comment (sarthak-amd) [reply tests/pytorch/triton_kernels/test_cast_mxfp4.py]: MXFP4 requires dimensions divisible by 32 for per-block scaling compatibility with AITER gemm_a4w4. I have added the shapes that should throw a valid and expected assertion error.

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py (hunk 4 above)
Comment (sarthak-amd) [reply tests/pytorch/triton_kernels/test_cast_mxfp4.py]: @wangye805 closely followed the example and updated the pytest.

**Comment by sudhu2k** [comment]
Hi @sarthak-amd 
Can't we merge the mxfp8 triton kernel with the mxfp4 triton kernel?https://github.com/ROCm/TransformerEngine/blob/ef83316a8f5de1a0bc6d1e574d385951f1de3194/transformer_engine/pytorch/triton_kernels/cast_transpose.py#L335

Kernel wise it should be exactly similar except how the block is being casted.
in mxfp8 we use a separate function to cast the values.
https://github.com/ROCm/TransformerEngine/blob/ef83316a8f5de1a0bc6d1e574d385951f1de3194/transformer_engine/pytorch/triton_kernels/cast_transpose.py#L301
So we can just put this part in a separate function:
https://github.com/ROCm/TransformerEngine/blob/ef83316a8f5de1a0bc6d1e574d385951f1de3194/transformer_engine/pytorch/triton_kernels/cast_transpose.py#L562-L571
and alternate the calls based on whether it's a MXFP8 kernel or MXFP4 kernel.


The following operation can be replaced with the [exp2f_rcp_triton function](https://github.com/ROCm/TransformerEngine/blob/ef83316a8f5de1a0bc6d1e574d385951f1de3194/transformer_engine/pytorch/triton_kernels/cast_transpose.py#L293C5-L293C21)

https://github.com/ROCm/TransformerEngine/blob/ef83316a8f5de1a0bc6d1e574d385951f1de3194/transformer_engine/pytorch/triton_kernels/cast_transpose.py#L549-L551

Another difference seems to be that MXFP8 kernels doesn't have the shuffle op fused in it. Do we want that mxfp8 as well? or is that specific to fp4 data?

Also @wangye805 noticed
For the column-wise quantization of an input tensor MxN, the mxfp8 seems to still have columnwise output as shape MxN but mxfp4 implementation is NxM. Technically we can add something like STORE_TRANSPOSE, to the kernel and invert the strides in the mxfp8 kernel itself. 

Let me know what you think @sarthak-amd

**Comment by sudhu2k** [comment]
## New changes

1. Refactor of kernel wrappers and mxfp4 tensors to be consistent with other recipes.
2. Decomposition of repeated work into a separate function for the mxfp4 quantize kernel.
3. Added shuffle tests.
4. Bug fixes.

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+        y1_scales_torch = ref_scale
+        if shuffle_B_matrix_for_aiter:
+            y1_scales_triton = un_shuffle_scales(
+                y1_scales_triton.view(y1_scales_triton.shape[0] // 32, -1)
+            )
+            y1_scales_torch = un_shuffle_scales(
+                y1_scales_torch.view(y1_scales_torch.shape[0] // 32, -1)
+            )
+        
+        compare_e8m0_scales(
```
Comment (wangye805) [inline tests/pytorch/triton_kernels/test_cast_mxfp4.py]: Okay, currently the target vs ref comparison is not coupled: we validated quantized data content and scale independently and allow for certain mismatch rate. 

A better way is to do the validation jointly by adjusting the quanized value if the scale is mismatched: https://github.com/ROCm/TransformerEngine/blob/9b312832f4a21bd5b73a6087de6bd1eeaf219b82/tests/cpp/operator/test_cast_mxfp8.cu#L314-L353

File: transformer_engine/common/util/pybind_helper.h (hunk 6 above)
Comment (wangye805) [reply transformer_engine/common/util/pybind_helper.h:112]: Since it's not changing the cpp part. We don't need those changes. Thanks

**Review by wangye805 — CHANGES_REQUESTED** [review]

File: transformer_engine/common/util/pybind_helper.h (hunk 6 above)
Comment (sudhu2k) [reply transformer_engine/common/util/pybind_helper.h:112]: We still need these changes because we have to expose it to python modules. https://github.com/ROCm/TransformerEngine/blob/4c5564a92bed816feb8b541f90f584dfefcc77f3/transformer_engine/pytorch/tensor/mxfp4_tensor.py#L56-L65

https://github.com/ROCm/TransformerEngine/blob/4c5564a92bed816feb8b541f90f584dfefcc77f3/transformer_engine/pytorch/tensor/mxfp4_tensor.py#L171

Comment (wangye805) [reply transformer_engine/common/util/pybind_helper.h:112]: Sorry about my confusion. I was saying we don't need other cpp code changes. Yeah, this change is needed

File: tests/pytorch/triton_kernels/test_cast_mxfp4.py (hunk 11 above)
Comment (sudhu2k) [reply tests/pytorch/triton_kernels/test_cast_mxfp4.py]: Added code to couple data and scale comparison by comparing dequantized results.

**Review by wangye805 — APPROVED** [review]
LGTM
